dmoj/judge.py

```diff
@@ -30,34 +36,53 @@ def setproctitle(title):
         pass
 
 
-class BatchBegin:
-    pass
+class IPC(Enum):
+    HELLO = 'HELLO'
+    BYE = 'BYE'
+    COMPILE_ERROR = 'COMPILE-ERROR'
+    COMPILE_MESSAGE = 'COMPILE-MESSAGE'
+    RESULT = 'RESULT'
+    BATCH_BEGIN = 'BATCH-BEGIN'
+    BATCH_END = 'BATCH-END'
+    GRADING_BEGIN = 'GRADING-BEGIN'
+    GRADING_END = 'GRADING-END'
+    GRADING_ABORTED = 'GRADING-ABORTED'
+    UNHANDLED_EXCEPTION = 'UNHANDLED-EXCEPTION'
+    REQUEST_ABORT = 'REQUEST-ABORT'
 
 
-class BatchEnd:
-    pass
+IPC_TEARDOWN_TIMEOUT = 5  # seconds
 
 
-class TerminateGrading(Exception):
-    pass
+logger = logging.getLogger(__name__)
 
 
-logger = logging.getLogger(__name__)
+Submission = NamedTuple(
+    'Submission',
+    [
+        ('id', int),
+        ('problem_id', str),
+        ('language', str),
+        ('source', str),
+        ('time_limit', float),
+        ('memory_limit', int),
+        ('short_circuit', bool),
+        ('meta', Dict),
+    ],
+)
 
 
 class Judge:
-    def __init__(self):
-        self.current_grader = None
-        self.current_submission_id = None
-        self.current_submission_thread = None
-        self._terminate_grading = False
-        self.packet_manager = None
+    def __init__(self, packet_manager: packet.PacketManager) -> None:
+        self.packet_manager = packet_manager
+        self.current_submission: Any = None
```

Comment:
Why not `Optional[Submission]`?

Reply:
Because then `packet` fails to type-check, since it accesses `judge.current_submission.id`, and I didn't want to `assert judge.current_submission` in _every single packet handler_.